### PR TITLE
[bugfix] check read_only when creating/dropping database

### LINKED ISSUE #2860: Execution of “create database system” succeeded
**Summary**

I tried to create a database with the same name as the system database, but it was successfully created. 

**How to reproduce**

**Error message and/or stacktrace**

image
image

[How to get the server logs](https://databend.rs/development/how-to-get-server-logs/)

**Additional context**

### PR DESCRIPTION
**Pull request by Veeupup**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

fix create duplicate database bug.

## Changelog

- Bug Fix

## Related Issues

Fixes #2860 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by BohuTANG** [comment]
Better to add a stateless test here, like:
https://github.com/datafuselabs/databend/blob/9c2e1d606688bbac5134eaaee33dfb3b798dfb8d/tests/suites/0_stateless/05_0001_ddl_create_database.sql#L7-L10

```
CREATE DATABASE system; -- {The ErrorCode } 
DROP DATABASE system; -- {The ErrorCode } 
```

File: query/src/catalogs/impls/catalog/overlaid_catalog.rs [hunk 1]
```diff
@@ -190,6 +190,19 @@ impl Catalog for OverlaidCatalog {
...
     ) -> common_exception::Result<CreateDatabaseReply> {
+        match self.read_only.get_database(&plan.db).await {
+            Ok(_) => {
+                return Err(ErrorCode::DatabaseAlreadyExists(format!(
+                    "{} database exists",
+                    plan.db
+                )));
+            }
+            Err(err) => {
+                if err.code() != ErrorCode::UnknownDatabase("").code() {
```
Comment (BohuTANG) [inline query/src/catalogs/impls/catalog/overlaid_catalog.rs:201]: Why we check the `UnknownDatabase` error code here not return the error?

Comment (BohuTANG) [reply query/src/catalogs/impls/catalog/overlaid_catalog.rs:201]: It seems we need a `self.read_only.exists_database` returns bool or true api here

Comment (Veeupup) [reply query/src/catalogs/impls/catalog/overlaid_catalog.rs:201]: when creating a DB, if the DB does not exist in read_only, it should be able to create in `bottom` and other errors should be return.

Comment (Veeupup) [reply query/src/catalogs/impls/catalog/overlaid_catalog.rs:201]: should I add the func `exists_database` in trait `Catalog`?

Comment (BohuTANG) [reply query/src/catalogs/impls/catalog/overlaid_catalog.rs:201]: Make sense.

**Comment by Veeupup** [comment]
@BohuTANG  I did not make trait func `exists_database` has default implementation because its signature will be
```rust
pub fn exists_database<'life0, 'life1, 'async_trait>(&'life0 self, db_name: &'life1 str) -> ::core::pin::Pin<Box<dyn ::core::future::Future<Output = Result<bool>> + ::core::marker::Send + 'async_trait>>
where
    'life0: 'async_trait,
    'life1: 'async_trait,
    Self: ::core::marker::Sync + 'async_trait
```
the signature has `Self`, and then Catalog will be not object safe.
So I have written some familiar code in different Catalog implements.

File: query/src/catalogs/impls/catalog/metastore_catalog.rs [hunk 2]
```diff
@@ -218,4 +218,17 @@ impl Catalog for MetaStoreCatalog {
 
         Ok(tbl)
     }
+
+    async fn exists_database(&self, db_name: &str) -> Result<bool> {
+        match self.get_database(db_name).await {
+            Ok(_) => Ok(true),
+            Err(err) => {
+                if err.code() == ErrorCode::UnknownDatabase("").code() {
```
Comment (BohuTANG) [inline query/src/catalogs/impls/catalog/metastore_catalog.rs:226]: ErrorCode::UnknownDatabase("").code() -> ErrorCode::UnknownDatabaseCode() is neat

File: query/src/catalogs/impls/catalog/system_catalog.rs [hunk 3]
```diff
@@ -177,4 +177,17 @@ impl Catalog for SystemCatalog {
 
         Ok(table.clone())
     }
+
+    async fn exists_database(&self, db_name: &str) -> Result<bool> {
+        match self.get_database(db_name).await {
+            Ok(_) => Ok(true),
+            Err(err) => {
+                if err.code() == ErrorCode::UnknownDatabase("").code() {
```
Comment (BohuTANG) [inline query/src/catalogs/impls/catalog/system_catalog.rs:185]: Same as above :)

File: query/src/catalogs/impls/catalog/metastore_catalog.rs (hunk 2 above)
Comment (Veeupup) [reply query/src/catalogs/impls/catalog/metastore_catalog.rs]: resolved

File: query/src/catalogs/impls/catalog/system_catalog.rs (hunk 3 above)
Comment (Veeupup) [reply query/src/catalogs/impls/catalog/system_catalog.rs]: resloved

**Pull request merged**